### PR TITLE
Strict per-treatment unreachable-reference check (#321 step 3)

### PR DESCRIPTION
**Pull request by JamesPHoughton**
## Summary

Third standalone piece of the #321 pipeline rewrite. New module `findUnreachableReferences.ts` implements the strict per-treatment reachable-keys check that the schema's existing `validateReferences` silently passes via its `globalProducedKeys` fallthrough.

Library-only, not wired into the editor yet (Phase 5). The diff orchestrator from #327 grows a new output bucket `unreachableReferences`.

## What the schema currently misses

For each reference in a treatment, the schema's checker walks down:

1. Producer in this treatment's `producedAt` → rank check
2. Producer in `laterPhaseKeys` → cross-phase forward-ref error
3. Producer in `globalProducedKeys` (anywhere in the file) → **silent pass**
4. Otherwise → unknown-reference error

Rung 3 masks three distinct runtime bugs that all look the same to a participant (the producer is never traversed):

- Cross-treatment leak (producer in another treatment)
- Producer in a template the consumer doesn't invoke
- Producer in a template another treatment invokes

## What this PR adds

`findUnreachableReferences(hydratedTreatmentFile)` does rung 3 strictly. It emits an issue when the reference's key is in `globalProducedKeys` but not in **this treatment's reachable set** (intros + this treatment's gameStages + this treatment's exit, with templates already expanded into those positions).

It doesn't duplicate the schema's existing diagnostics:
- Pure typos (key produced nowhere) → schema handles
- Forward references within a treatment → schema's rank check handles
- Cross-phase forward refs → schema's existing check handles
- The strict rung-3 case → **this function**

## Single diagnostic class, not three

I considered three distinct messages (cross-treatment vs uninvoked-template vs nowhere-at-all) but they'd be prescriptive in ways that risk mis-directing the fix. One message says "this reference doesn't match any element in this treatment" plus the storage key — the user has the whole picture and can investigate without the validator guessing.

## Only sound on hydrated form

Per-treatment `producedAt` on raw source is incomplete (templates not expanded), so applying strict rules pre-fill would false-positive on legitimate template-injected references. The function should only be called on the hydrated form, which `runValidationDiff` already produces.

## Touchpoints

- [`validateReferences.ts`](packages/stagebook/src/schemas/validateReferences.ts): export `RefSite`, `STAGE_PRODUCED_REF_TYPES`, `enumerateStepSites`, `collectStepKeys`, `collectKeysFromAny`, `collectProducedKeys` for reuse. **Schema behavior unchanged** — existing reference check still runs with fallthrough, just like before.
- [`findUnreachableReferences.ts`](packages/stagebook/src/schemas/findUnreachableReferences.ts) (new): the strict check + 9 unit tests
- [`runValidationDiff.ts`](packages/stagebook/src/schemas/runValidationDiff.ts): add `unreachableReferences` to `ValidationDiffResult`; orchestrator calls the new function after hydration
- [`runValidationDiff.test.ts`](packages/stagebook/src/schemas/runValidationDiff.test.ts): 2 new tests demonstrating the bucket
- `schemas/index.ts`: export the new function

## Test plan

- [x] 9 new tests in `findUnreachableReferences.test.ts` covering happy path, cross-treatment leak, same-treatment earlier-stage (no false positive), intro producer (no false positive), external sources (no false positive), schema-typo-not-duplicated, multi-treatment per-scope, and malformed input edge cases
- [x] 2 new tests in `runValidationDiff.test.ts` showing the orchestrator routes leaks into `unreachableReferences` (not source/hydrated/matched/sourceOnly/hydratedOnly) and returns empty on hydration failure
- [x] All 974 stagebook + 87 viewer + 201 vscode tests pass
- [x] Lint clean, format clean

## Relationship to #321

Step 3 of the eight-step pipeline. Builds on #324 (pre-hydration semantic) and #327 (diff orchestrator). Next PR (Phase 5) wires the full orchestrator — including this new bucket — into the editor's async `validateTreatmentFile`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/stagebook/src/schemas/findUnreachableReferences.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  const treatments = toArray(hydratedTreatmentFile.treatments);
+  const templates = toArray(hydratedTreatmentFile.templates);
+
+  // Intro keys are shared across treatments (every treatment can read
+  // intro-phase data). Build once.
+  const introProducedKeys = new Set<string>();
+  for (const seq of introSequences) {
+    if (!isRecord(seq)) continue;
+    for (const step of toArray(seq.introSteps)) {
+      collectProducedKeys(step, introProducedKeys);
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply packages/stagebook/src/schemas/findUnreachableReferences.ts]: Confirmed bug, fixed in 0b23ebb. `collectProducedKeys` checks `typeof el.type === "string"` and skips when it's not an element — passing a step object made `introProducedKeys` empty. The existing intro-producer tests passed only by accident: `globalProducedKeys` was also missing intro keys, so the function hit the "key produced nowhere → schema handles" early-out instead of the correct reachable path. Switched to `collectStepKeys(step)` which iterates `step.elements`. The intro-producer test now exercises the right code path.

File: packages/stagebook/src/schemas/runValidationDiff.ts [hunk 2]
```diff
@@ -249,13 +263,19 @@ export function runValidationDiff({
     hydratedIssues,
   );
 
+  // Strict per-treatment reachable-keys check, only sound on the
+  // hydrated form. Catches what the schema's existing reference
+  // checker silently passes via globalProducedKeys fallthrough.
+  const unreachableReferences = findUnreachableReferences(expanded);
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply packages/stagebook/src/schemas/runValidationDiff.ts]: Confirmed — `fillTemplates` strips `templates:` from its output, so reading `hydratedTreatmentFile.templates` returns nothing. Fixed in 0b23ebb by adding an explicit `options.templates` parameter; `runValidationDiff` now passes the merged template list (root + imports) through. The orchestrator was already computing `mergedTemplates` so it's a one-line forward in the call site.

File: packages/stagebook/src/schemas/findUnreachableReferences.test.ts [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+ * All three look the same at runtime — the consumer's participant
+ * never traverses the producer — and all three produce the same
+ * diagnostic class ("not reachable from this treatment"). Three
+ * sub-cases collapsed into one message; users don't need a
+ * prescriptive hint to find the bug.
+ *
+ * Caveat: only sound on the HYDRATED form, where each treatment's
+ * `producedAt` is complete. Running on raw source would false-
+ * positive on legitimate template-injected references.
+ */
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply packages/stagebook/src/schemas/findUnreachableReferences.test.ts:24]: Added in 0b23ebb. Two tests:

- `flags a reference to a key produced only by a template that no treatment invokes` — passes the merged templates via the new option, expects the leak to surface
- `without passing templates, an uninvoked-template producer's key looks like a typo and is left to the schema` — documents the contract: callers that want full coverage must pass templates, otherwise the function defers to the schema's unknown-reference check

Combined with the intro-producer fix from comment 1, this is a substantial correctness improvement — the function previously failed silently for the uninvoked-template case while pretending to handle it.
